### PR TITLE
feat: bosch smoke intruder and smoke alarm triggers

### PR DESCRIPTION
**Pull request by vjonas**
I ran into formatting issues. It logs too many changes. Not sure how to align that, how should I do that? @Koenkk
Potentially we could enable some formatter as Prettier to the codebase?
This should allow users to send a mqtt topic to toggle the alarm and intruder state of the smoke detector II.

Add these 2 commands to trigger the intruder and alarm state:
image

### CODE REVIEW

**Comment by Koenkk** [comment]
Please also undo the formatting changes here.

**Comment by vjonas** [comment]
> Please also undo the formatting changes here.

Should be good now.

File: src/devices/bosch.ts [hunk 1]
```diff
@@ -908,14 +917,90 @@ const definitions: Definition[] = [
...
-        toZigbee: [],
+        toZigbee: [{
+            key: ['intruder_alarm_state', 'smoke_alarm_state', 'boschSmokeDetectorSiren', 'command'],
+            convertSet: async (entity, key, value:smokeSensorData, meta) => {
+                const dataToSend = {
+                    cluster: value?.cluster,
+                    command: value?.command,
+                    payload: value?.payload,
+                };
+                if (key === 'smoke_alarm_state' || key==='intruder_alarm_state') {
```
Comment (Koenkk) [inline src/devices/bosch.ts:928]: What's the difference between `smoke_alarm_state` and `intruder_alarm_state`?

File: src/devices/bosch.ts [hunk 2]
```diff
@@ -908,14 +917,90 @@ const definitions: Definition[] = [
         vendor: 'Bosch',
         description: 'Smoke alarm detector',
         fromZigbee: [fz.battery, fz.ias_smoke_alarm_1],
-        toZigbee: [],
+        toZigbee: [{
+            key: ['intruder_alarm_state', 'smoke_alarm_state', 'boschSmokeDetectorSiren', 'command'],
```
Comment (Koenkk) [inline src/devices/bosch.ts:921]: Why is `'boschSmokeDetectorSiren', 'command'` needed here?

File: src/devices/bosch.ts (hunk 1 above)
Comment (vjonas) [reply src/devices/bosch.ts:928]: Smoke alarm is triggering a different sound pattern than intruder alarm. Separated them to have 2 distinct toggles.

File: src/devices/bosch.ts (hunk 2 above)
Comment (vjonas) [reply src/devices/bosch.ts]: I believe it is not necessary. I'll remove it and re check.

**Comment by vjonas** [comment]
I removed the unnecessarily commands and pushed.

**Comment by vjonas** [comment]
Tests passed too now.

File: src/devices/bosch.ts [hunk 3]
```diff
@@ -908,14 +917,90 @@ const definitions: Definition[] = [
...
+            device.save();
+            await endpoint.unbind('genPollCtrl', coordinatorEndpoint);
         },
-        exposes: [e.smoke(), e.battery(), e.battery_low(), e.test()],
+        exposes: [e.smoke(), e.battery(), e.battery_low(), e.test(),
+            e
+                .binary(
+                    'intruder_alarm_state',
+                    ea.ALL,
+                    smokeDetectorAlarmState.intruder_on.toString(),
```
Comment (Koenkk) [inline src/devices/bosch.ts:992]: I think it's better to just have `ON`/`OFF` here and do the conversion to a number in the toZigbeeConverter

File: src/devices/bosch.ts [hunk 4]
```diff
@@ -908,14 +917,90 @@ const definitions: Definition[] = [
...
+                'ssIasWd',
+                'genBasic',
+                64684]);
             await reporting.batteryPercentageRemaining(endpoint);
+            device.defaultSendRequestWhen = 'immediate';
+            device.save();
+            await endpoint.unbind('genPollCtrl', coordinatorEndpoint);
         },
-        exposes: [e.smoke(), e.battery(), e.battery_low(), e.test()],
+        exposes: [e.smoke(), e.battery(), e.battery_low(), e.test(),
```
Comment (Koenkk) [inline src/devices/bosch.ts:987]: Is it possible to turn on both smoke and intruder sound on at the same time? If not I think it makes more sense to have 1 expose. (an `e.enum()`)

Comment (vjonas) [reply src/devices/bosch.ts:987]: Yes it is possible. The sound gets queued.
Ex: turning on smoke and intruder. Turning off smoke results in the intruder alarm being played. They both require a different 'OFF' value to be send.

To turn of smoke, value **46080** needs to be send, and for intruder value: **46081** needs to be send.

I believe the enum falls short here. The convertSet could send out 2 commands but that feels hacky too.

What is your suggestion? @Koenkk

Comment (Koenkk) [reply src/devices/bosch.ts:988]: Then we should keep it as is (2 separate exposes)

File: src/devices/bosch.ts (hunk 3 above)
Comment (Koenkk) [reply src/devices/bosch.ts:993]: @vjonas can you check this comment?

File: src/devices/bosch.ts (hunk 4 above)
Comment (vjonas) [reply src/devices/bosch.ts:987]: Ok ready for merge I believe.

File: src/devices/bosch.ts (hunk 3 above)
Comment (vjonas) [reply src/devices/bosch.ts:992]: Oh yes sorry I overlooked this one. Sure!

Comment (vjonas) [reply src/devices/bosch.ts]: Should be good now.

**Comment by Koenkk** [comment]
Thanks!

**Comment by N3rdix** [comment]
Thank you so much @vjonas, awesome work ❤️ 

Just tested the state triggers and they work fine. Getting the state of the trigger throws these errors for me when I press the refresh button, is this to be expected?

```
Error 2024-01-16 13:45:20Publish 'get' 'smoke_alarm_state' to 'Rauchmelder Test' failed: 'Error: Read 0xb43a31fffe872553/1 ssIasZone([240], {"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":4617,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Status 'UNSUPPORTED_ATTRIBUTE')'
Error 2024-01-16 13:45:22Publish 'get' 'intruder_alarm_state' to 'Rauchmelder Test' failed: 'Error: Read 0xb43a31fffe872553/1 ssIasZone([240], {"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":4617,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Status 'UNSUPPORTED_ATTRIBUTE')'
```

image

**Comment by vjonas** [comment]
> Thank you so much @vjonas, awesome work ❤️
> 
> Just tested the state triggers and they work fine. Getting the state of the trigger throws these errors for me when I press the refresh button, is this to be expected?
> 
> ```
> Error 2024-01-16 13:45:20Publish 'get' 'smoke_alarm_state' to 'Rauchmelder Test' failed: 'Error: Read 0xb43a31fffe872553/1 ssIasZone([240], {"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":4617,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Status 'UNSUPPORTED_ATTRIBUTE')'
> Error 2024-01-16 13:45:22Publish 'get' 'intruder_alarm_state' to 'Rauchmelder Test' failed: 'Error: Read 0xb43a31fffe872553/1 ssIasZone([240], {"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":4617,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Status 'UNSUPPORTED_ATTRIBUTE')'
> ```
> 
> image

Glad to hear!

I believe so, I am not sure yet how that refresh mechanism works. Maybe something to sort out later.

**Pull request merged**